output/badger/output.go

```diff
@@ -0,0 +1,202 @@
+	}
+
+	return octoWatermark.AsTime(), nil
+}
+
+func (o *Output) MarkEndOfStream(ctx context.Context, tx storage.StateTransaction) error {
+	endOfStreamState := storage.NewValueState(tx.WithPrefix(endOfStreamPrefix))
+
+	phantom := octosql.MakePhantom()
+	if err := endOfStreamState.Set(&phantom); err != nil {
+		return errors.Wrap(err, "couldn't mark end of stream")
+	}
+
+	return nil
+}
+
+func (o *Output) GetEndOfStream(ctx context.Context, tx storage.StateTransaction) (bool, error) {
+	endOfStreamState := storage.NewValueState(tx.WithPrefix(endOfStreamPrefix))
+
+	var octoEndOfStream octosql.Value
+	err := endOfStreamState.Get(&octoEndOfStream)
+	if err == storage.ErrNotFound {
+		return false, nil
+	} else if err != nil {
+		return false, errors.Wrap(err, "couldn't get end of stream value")
+	}
+
+	return true, nil
+}
+
+func (o *Output) MarkError(ctx context.Context, tx storage.StateTransaction, err error) error {
+	endOfStreamState := storage.NewValueState(tx.WithPrefix(errorPrefix))
+
+	octoError := octosql.MakeString(err.Error())
+	if err := endOfStreamState.Set(&octoError); err != nil {
+		return errors.Wrap(err, "couldn't mark error")
+	}
+
+	return nil
+}
+
+func (o *Output) GetErrorMessage(ctx context.Context, tx storage.StateTransaction) (string, error) {
+	errorState := storage.NewValueState(tx.WithPrefix(errorPrefix))
+
+	var octoError octosql.Value
+	err := errorState.Get(&octoError)
+	if err == storage.ErrNotFound {
+		return "", nil
+	} else if err != nil {
+		return "", errors.Wrap(err, "couldn't get error message")
+	}
+
+	return octoError.AsString(), nil
+}
+
+func (o *Output) Close() error {
+	return nil // TODO: Cleanup?
+}
+
+func (o *Output) ListRecords(ctx context.Context, tx storage.StateTransaction) ([]*execution.Record, error) {
+	records := storage.NewMap(tx.WithPrefix(recordsPrefix))
+
+	iter := records.GetIterator()
+
+	var outRecords []*execution.Record
+	var err error
+	var octoKey octosql.Value
+	var octoIDs octosql.Value
+	for err = iter.Next(&octoKey, &octoIDs); err == nil; err = iter.Next(&octoKey, &octoIDs) {
+		object := octoKey.AsMap()
+		var fields []string
+		for k := range object {
+			fields = append(fields, k)
+		}
+		sort.Strings(fields)
+
+		data := make([]octosql.Value, len(fields))
+		for i := range fields {
+			data[i] = object[fields[i]]
+		}
+
+		variableNames := make([]octosql.VariableName, len(fields))
+		for i := range fields {
+			variableNames[i] = octosql.NewVariableName(fields[i])
+		}
+
+		for _, id := range octoIDs.AsSlice() {
+			outRecords = append(
+				outRecords,
+				execution.NewRecordFromSlice(
+					variableNames,
+					data,
+					execution.WithID(execution.NewRecordID(id.AsString())),
+					execution.WithEventTimeField(o.EventTimeField),
+				),
+			)
+		}
+	}
+	if err := iter.Close(); err != nil {
+		return nil, errors.Wrap(err, "couldn't close iterator")
+	}
+	if err != storage.ErrEndOfIterator {
```

Comment:
Shouldn't this be directly after for loop? It probably doesn't matter but conceptually first we want to check if the for loop finished without an error and then close the iterator

Reply:
sure